coordinator/src/main/scala/filodb.coordinator/queryengine2/QueryEngine.scala

```diff
@@ -203,10 +203,22 @@ class QueryEngine(dataset: Dataset,
      *
      * Starting off with solution 1 first until (2) or some other approach is decided on.
      */
-    toReduce.plans.foreach(_.addRangeVectorTransformer(AggregateMapReduce(lp.operator, lp.params, lp.without, lp.by)))
-    // One could do another level of aggregation per node too. Ignoring for now
-    val reduceDispatcher = pickDispatcher(toReduce.plans)
-    val reducer = ReduceAggregateExec(queryId, reduceDispatcher, toReduce.plans, lp.operator, lp.params)
+    toReduceLevel1.plans.foreach(
+      _.addRangeVectorTransformer(AggregateMapReduce(lp.operator, lp.params, lp.without, lp.by))
+    )
+
+    val toReduceLevel2 =
+      if (toReduceLevel1.plans.size >= 16) {
+        // If number of children is above a threshold, parallelize aggregation
+        val groupSize = Math.sqrt(toReduceLevel1.plans.size).ceil.toInt
```

Comment:
How was sqrt determined to be a good function?

Reply:
Right now it is just hypothesis, to perform equal number of reductions at every node, thereby distributing work equally. We would need to test this hypothesis in a real environment. Network latency is also at play here.